subprojects/core/src/test/groovy/org/gradle/cache/internal/DefaultCacheAccessTest.groovy

```diff
@@ -631,4 +634,119 @@ class DefaultCacheAccessTest extends ConcurrentSpec {
+
+        cleanup:
+        access?.close()
+    }
+
+    def "throws InvalidCacheReuseException when cache value type differs"() {
+        def access = newAccess(None)
+
+        when:
+        access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class))
+        access.newCache(new PersistentIndexedCacheParameters('cache', String.class, String.class))
+
+        then:
+        thrown(DefaultCacheAccess.InvalidCacheReuseException)
+
+        cleanup:
+        access?.close()
+    }
+
+    def "throws InvalidCacheReuseException when cache key type differs"() {
+        def access = newAccess(None)
+
+        when:
+        access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class))
+        access.newCache(new PersistentIndexedCacheParameters('cache', Integer.class, Integer.class))
+
+        then:
+        thrown(DefaultCacheAccess.InvalidCacheReuseException)
+
+        cleanup:
+        access?.close()
+    }
+
+    def "throws InvalidCacheReuseException when cache decorator differs"() {
+        def access = newAccess(None)
+        def decorator = Mock(CacheDecorator)
+        lockManager.lock(lockFile, mode(Exclusive), "<display-name>") >> lock
+        decorator.decorate(_, _, _, _, _) >> { String cacheId, String cacheName, MultiProcessSafePersistentIndexedCache persistentCache, CrossProcessCacheAccess crossProcessCacheAccess, AsyncCacheAccess asyncCacheAccess ->
+            persistentCache
+        }
+
+        when:
+        access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class))
+        access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class).cacheDecorator(decorator))
+
+        then:
+        thrown(DefaultCacheAccess.InvalidCacheReuseException)
+
+        cleanup:
+        access?.close()
+    }
+
+    def "returns the same cache object when cache decorator match"() {
+        def access = newAccess(None)
+        def decorator1 = Mock(CacheDecorator)
+        def decorator2 = Mock(CacheDecorator)
+        lockManager.lock(lockFile, mode(Exclusive), "<display-name>") >> lock
+        decorator1.decorate(_, _, _, _, _) >> { String cacheId, String cacheName, MultiProcessSafePersistentIndexedCache persistentCache, CrossProcessCacheAccess crossProcessCacheAccess, AsyncCacheAccess asyncCacheAccess ->
+            persistentCache
+        }
+        decorator2.decorate(_, _, _, _, _) >> { String cacheId, String cacheName, MultiProcessSafePersistentIndexedCache persistentCache, CrossProcessCacheAccess crossProcessCacheAccess, AsyncCacheAccess asyncCacheAccess ->
+            persistentCache
+        }
+
+        when:
+        def cache1 = access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class).cacheDecorator(decorator1))
+        def cache2 = access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class).cacheDecorator(decorator2))
```

Comment:
Looking closer at the code, I see that the only `CacheDecorator` implementation available is owned by the global scope. This means there will only be one instance. This test support for different instance. The check that was added only checks for a matching class name. Implementing [`InMemoryTaskArtifactCache.equals`](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/InMemoryTaskArtifactCache.java) doesn't look trivial because of the internal state. The best equality would probably be matching the instance which would break this test. What would make the most sense:
1. Match the instance only (very strict)
2. Match the class name (relaxed but may hide a bug)
3. Provide a full `equals` implementation (not trivial if even possible)